OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/SSHAgentService.java

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (C) 2017 Andreas-Christian Hagau <dev@hagau.se>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.sufficientlysecure.keychain.remote;
+
+import android.app.PendingIntent;
+import android.app.Service;
+import android.content.Intent;
+import android.os.IBinder;
+import android.util.Log;
+
+import org.bouncycastle.bcpg.HashAlgorithmTags;
+import org.openintents.ssh.ISSHAgentService;
+import org.openintents.ssh.SSHAgentApiConstants;
+import org.sufficientlysecure.keychain.Constants;
+import org.sufficientlysecure.keychain.pgp.CanonicalizedPublicKey;
+import org.sufficientlysecure.keychain.pgp.CanonicalizedPublicKeyRing;
+import org.sufficientlysecure.keychain.pgp.exception.PgpGeneralException;
+import org.sufficientlysecure.keychain.ssh.AuthenticationData;
+import org.sufficientlysecure.keychain.ssh.AuthenticationOperation;
+import org.sufficientlysecure.keychain.operations.results.OperationResult.LogEntryParcel;
+import org.sufficientlysecure.keychain.ssh.AuthenticationResult;
+import org.sufficientlysecure.keychain.pgp.exception.PgpKeyNotFoundException;
+import org.sufficientlysecure.keychain.provider.ApiDataAccessObject;
+import org.sufficientlysecure.keychain.provider.CachedPublicKeyRing;
+import org.sufficientlysecure.keychain.provider.KeyRepository;
+import org.sufficientlysecure.keychain.provider.KeychainContract;
+import org.sufficientlysecure.keychain.service.input.CryptoInputParcel;
+import org.sufficientlysecure.keychain.service.input.RequiredInputParcel;
+import org.sufficientlysecure.keychain.ssh.AuthenticationParcel;
+
+import java.security.PublicKey;
+import java.util.Date;
+import java.util.HashSet;
+
+
+public class SSHAgentService extends Service {
```

Comment:
Regarding the naming: I would propose will call this ``SSHService``

Reply:
Or better: ``SshService``

Reply:
Well, the service is not doing ssh, it's doing ssh agent or ssh auth stuff. I would propose SshAgentService or SshAuthService

Reply:
I think agent is the term that is used on desktop system. No need to use it in our implementation. The API itself is the SSH API living in the org.openintents.ssh package, so that is why I propose to simply call it SshService because it is implementing this SSH API. Maybe in the future it could do more than just authentication, we never know.

Reply:
Hmm, maybe I see your pont but then we should rename the whole API.

Reply:
I don't see how desktop/mobile makes a difference. The purpose of naming is to tell the reader what something does, and "agent" is a perfectly established term for the exact thing this service is doing.

I see the argument about calling it "SshService" because it implements the "SshApi", but my conclusion there would be that "SshApi" is also not a well chosen name, for the same reasons.

Reply:
I would be okay with ``SshAuthenticationService``, but then I would also rename the package to ``org.openintents.sshauthentication``.

Reply:
sounds good to me :+1: (ftr, I missed your second reply while writing that last one)

Reply:
I moving the package to org.openintents.ssh.authentication makes sense regardless of other naming issues (btw. while doing some cleaning, I've been renaming classes and methods from *auth* to *authentication*, local variables can stay *auth*)

Reply:
Great! I am not sure about ``org.openintents.ssh.authentication`` in comparison to ``org.openintents.sshauthentication``. What do you think @Valodim ?

Reply:
`.ssh.authentication` sounds good to me

Reply:
That's how it should be!

Reply:
SshAuthenticationService makes sense. Renaming the API to sshauthentication-api and SshAuthenticationApi seems unwieldy, but would be logical.
There is of course nothing really specific about SSH to the API (at least right now), really it is just a challenge-response authentication API...

Reply:
Hmh, I actually like ``SshAuthenticationService``, ``sshauthentication-api`` and ``SshAuthenticationApi``.

As said before I would prefer SSH in the name to make the purpose clear for an outsider, like SSH client developers.

Reply:
I agree. Also, googleability of the term.

Reply:
SshAuthenticationService it is, then. For the API itself, I don't know yet :)